Better default styles computation.

IDisposable opened the pull request:
Uses the correct tag hierarchy when computing the default style to account for (implicit) browser stylesheets that use tags within a hierarchy to specify default styles.

We create the elements in the sandbox with all the required parent nodes to match up to the first [block-level element](https://developer.mozilla.org/en-US/docs/Web/HTML/Block-level_elements), compute the default styling and then tear down that hierarchy. This work is still only done once per tag-and-parents combination.

Resolves #106 and thus also Resolves #95

IDisposable committed:
Extend timeout of table test

IDisposable committed:
Mirror the cloned-element's DOM fragment hierarchy
This ensures that browser-stylesheets that have parent tag based rules
work correctly. This is evident when doing (for example) a bare `TD` before, the default style was defaulted to be `padding: opx`, but when the `TD` is actually created as a child of the normal hierarchy (e.g. `TABLE`, `TBODY`, `TR`, `TD`) then the defaulted padding is `padding: 1px`.
Refactored a bit for clarity
Resolves #106 and thus also Resolves #95

IDisposable commented:
Hey @meche-gh, this should do the trick... I also refactored that part because it was getting a bit long-winded.

IDisposable committed:
Prettify and build

meche-gh commented on src/dom-to-image-more.js:
```diff
@@ -1010,45 +1012,147 @@
+
+    function getDefaultStyle(sourceElement) {
+        const tagHierarchy = computeTagHierarchy(sourceElement);
+        const tagKey = tagHierarchy.join('>'); // it's like CSS
```
This adds a cost for generalised tags like span and i which I suspect is exponential - $O(c^N)$ where $c$ is average hierarchy depth for that tag. As we're constructing against tag hierarchy in the DOM, where most developers religiously follow "permitted (non-normative) context", we should be able to get away with caching by tag name, or reducing the key to a maximum of 2 - `tagName` or `ascentStopper tagName`. Need to protect against #70 regression here

meche-gh replied:
The ideal way to stress this algo in performance testing is to go to a complex Wikipedia article like https://en.wikipedia.org/wiki/JavaScript?useskin=vector-2022

If you have qualms about overly simplifying the caching logic, we can loop in the original author and see what they think RE: perf cost-benefit

IDisposable replied:
Yep, that would do it _IFF_ the assertion is that `TD>TR>TBODY>TABLE` _results_ in the same default styles as `TD>TR>TFOOT>TABLE` and other similar, but not sufficiently dissimilar fragments (both would end up with the `tagKey` of `'TD>TABLE'`. I'm going to see what it looks like. Do you have a performance test-jig for doing arbitrary pages? Might be a nice jsfiddle

IDisposable replied:
I've updated for the option to be `strict` or `relaxed` in case someone breaks. We default to `strict`

IDisposable commented:
That's a fair concern... I ran some tests against some huge tests I have
and found 5 tag depth is pretty much the max before hitting a stopper since
we're not factoring siblings in (and siblings are a cache hit). I suspect
it's not a huge issue, but we would need at least 4 for the fix to kick in
for `TD` cells.

I wonder if we have a reproduction for the #70 issue handy?

Also wondering if we would be better served for now in pulling in your post
filter stuff and maybe putting the optimizations behind a feature flag?

meche-gh commented:
For `td` - if we cache as `tagName` or `acsentStopper tagName` - the fix should always kick in when TD is wrapped by TABLE, and we get new cache entries for tag abuse e.g. div>td. Plus tags like SPAN and I will probably only produce the permutations `span, div span, p span`

IDisposable commented:
That may work... we're not going to save anything walking up the parentage,
and we need to create the tags we found. So this would effectively save
caching a few edge cases (and smaller tagKey strings...but that is a tiny
factor). I will test with a bunch of big pages and see what an instrumented
run shows.

meche-gh commented:
> Also wondering if we would be better served for now in pulling in your post
filter stuff and maybe putting the optimizations behind a feature flag?

I was thinking the filter could be used for the other dom-to-image fork packages. At the same time, I'd imagine a lot of opt-ins. I'll see if I can refactor my secret WIP branch into a neat minimal PR.

meche-gh commented:
Just to be clear, we do want to be creating keys against the hierarchy tree, and my concern about the perf can be handled by a change to the logic for creating the cache key.

meche-gh commented on src/dom-to-image-more.js:
```diff
@@ -1010,45 +1012,147 @@
+        'P',
+        'PRE',
+        'SECTION',
+        'SVG',
+        'TABLE',
+        'UL',
+        // these are ultimate stoppers in case something drastic changes in how the DOM works
+        'BODY',
+        'HEAD',
+        'HTML',
```
[nit] The `svg` addition is ultimately custom - it's a performance boost!
SVG tags inside an `<svg>` are not going to have their default CSS change because of the parent elements outside of their `<svg>` tag:

```suggestion
        'TABLE',
        'UL',
        // these are ultimate stoppers in case something drastic changes in how the DOM works
        'BODY',
        'HEAD',
        'HTML',
        'SVG',
```

meche-gh replied:
You can also add `<math>` as the MathML sectioning root. It is not part of the HTML spec unfortunately, but web browsers irregularly implement it as a separate concurrent web technology

IDisposable replied:
`math` added, in lower case... I suspect when browsers actually do adopt the tag it'll morph to uppercase...

meche-gh replied:
Cool. We need to put `svg` (another lowercase exception) after the "these are ultimate stoppers" comment.  
Just to be sure it doesn't get deleted later when the list gets updated

meche-gh replied:
- We need to lowercase `svg` also!
- The `svg` tag won't appear in the MDN list so we should move it under the `// these are ultimate stoppers` comment

IDisposable replied:
Oh, doh... I can fix that real quick... completely missed what you were going for...

IDisposable committed:
Add math test

IDisposable committed:
Fix image comparison in test so you get the Base64

IDisposable committed:
Added a test for styles computation

Also cleaned up the Promise handling to return the right things along the chain.

IDisposable committed:
Added support for strict or relaxed style cache keys

This necessitates passing the entire `options` block down the `cloneNode` path, that will be handy later anyway.
Moved `SVG` tag to where it makes sense.

IDisposable committed:
Add `math` tag to ascentStoppers.

IDisposable committed:
Reenable the rests of the tests

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -1010,45 +1012,147 @@
         return defaultStyle;
+
+        function computeTagHierarchy(sourceNode) {
+            const ELEMENT_NODE = 1;
```
Pulled those changes in and now have the ability to key the defaultStyleCache with either the fill tag hierachy (`strict` mode) or just `ascentStopper + tagName` ('relaxed` mode). This should make it possible revert if we broke something.
